sfge/src/main/java/com/salesforce/rules/UnusedMethodRule.java

```diff
@@ -161,84 +200,204 @@ private boolean methodIsIneligible(MethodVertex vertex) {
+     */
+    private boolean methodUsedBySuperclass(MethodVertex methodVertex) {
+        // TODO: IMPLEMENT THIS METHOD
+        return false;
+    }
+
+    /**
+     * @return - True if the provided method is visible to inner classes of the class where it's
+     *     defined. E.g., if it's a static method.
+     */
+    private boolean methodVisibleToInnerClasses(MethodVertex methodVertex) {
+        // TODO: IMPLEMENT THIS METHOD
+        return false;
+    }
+
+    /**
+     * Seeks invocations of the provided method by inner classes of its host class.
+     *
+     * @return - True if such an invocation is found, else false.
+     */
+    private boolean methodUsedByInnerClass(MethodVertex methodVertex) {
+        // TODO: IMPLEMENT THIS METHOD
+        return false;
+    }
+
+    /** @return - True if the provided method is visible externally, e.g. it's a public method. */
+    private boolean methodVisibleExternally(MethodVertex methodVertex) {
+        // TODO: IMPLEMENT THIS METHOD
+        return false;
+    }
+
+    private boolean methodUsedExternally(MethodVertex methodVertex) {
+        // TODO: IMPLEMENT THIS METHOD
+        return false;
+    }
+
+    /**
+     * Helper method for {@link #methodIsIneligible(MethodVertex)}. Indicates whether a method is
+     * annotated with an engine directive denoting that this rule should skip it.
+     */
+    private boolean directedToSkip(MethodVertex methodVertex) {
+        List<EngineDirective> directives = methodVertex.getAllEngineDirectives();
+        for (EngineDirective directive : directives) {
+            if (directive.isAnyDisable()
+                    && directive.matchesRule(this.getClass().getSimpleName())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /** Returns a list of every method call occurring in the specified class. */
+    private List<InvocableWithParametersVertex> getMethodCalls(String definingType) {
         // First, check if we've got anything for the desired type.
         // If so, we can just return that.
-        if (this.internalMethodCallsByDefiningType.containsKey(definingType)) {
-            return this.internalMethodCallsByDefiningType.get(definingType);
+        if (this.methodCallsByDefiningType.containsKey(definingType)) {
+            return this.methodCallsByDefiningType.get(definingType);
         }
-        // Otherwise, we'll need to do a query for internal calls.
-        // For instance methods, an internal call is any method call whose reference is `this` or
-        // empty.
-        String[] desiredReferenceTypes =
-                new String[] {
-                    NodeType.THIS_VARIABLE_EXPRESSION, NodeType.EMPTY_REFERENCE_EXPRESSION
-                };
-        List<InvocableWithParametersVertex> internalCalls =
+        // Otherwise, we'll need to do a query.
+        // Any node with one of these labels is a method call.
+        List<String> targetLabels =
```

Comment:
This method now gets and caches every single method call in the class, instead of just the ones that look like internal calls. It's now on the visitors to ignore method calls that don't match their expectations.